pytorch_lightning/trainer/supporters.py

```diff
@@ -371,6 +363,56 @@ def __init__(self, loaders: Any, mode: str = 'min_size'):
         if self.mode == 'max_size_cycle':
             self._wrap_loaders_max_size_cycle()
 
+        self.loaders_iter_state_dict = None
+
+    def state_dict(self, num_batches_processed: int):
+        if not fault_tolerant_enabled():
+            return {}
+
+        def state_dict_fn(dataloader: DataLoader, iterator: Iterator) -> Dict:
+            nonlocal num_batches_processed
+            # find next worker if multiple workers were used
+            state = find_current_worker(iterator)
+            if isinstance(dataloader.dataset, CaptureIterableDataset):
+                # the sampler state dict are extracted in ``CombinedLoaderIterator``
+                if iterator is not None and getattr(iterator, "_sampler_state_dict", None) is not None:
+                    state.update(iterator._sampler_state_dict[0])
+            else:
+                # fetch directly from fast forward sampler
+                state.update(dataloader.fast_forward_sampler.state_dict(num_batches_processed))
+            return DataLoaderDict(state)
+
+        return apply_to_collections(self.loaders, self._iterator.loader_iters, (Iterator, DataLoader), state_dict_fn)
+
+    def load_state_dict(self, state_dict):
+        self.loaders_iter_state_dict = state_dict
+
+        def mock_reset_fn(self, *_, **__):
+            pass
+
+        # delay reset call.
+        _MultiProcessingDataLoaderIter._ori_reset = _MultiProcessingDataLoaderIter._reset
+        _MultiProcessingDataLoaderIter._reset = mock_reset_fn
+
+    def on_restart(self, iterator: Iterator):
+        if self.loaders_iter_state_dict:
+
+            def create_loader_iters(dataloader: DataLoader, state_dict: DataLoaderDict):
+                if isinstance(dataloader.dataset, CaptureIterableDataset):
+                    dataloader.dataset.load_state_dict(state_dict)
+                else:
+                    dataloader.fast_forward_sampler.load_state_dict(state_dict)
+                iterator = cycle_to_next_worker_and_reset(dataloader, state_dict)
+                if isinstance(dataloader.dataset, CaptureIterableDataset):
+                    state_dict = {k: v for k, v in state_dict.items() if k not in ("num_worker", "previous_worker")}
+                    # need to re-attach the ``state dict`` into the iterator for future collection.
+                    iterator._sampler_state_dict = [state_dict]
+                return iterator
+
+            iterator._loader_iters = apply_to_collections(
+                self.loaders, self.loaders_iter_state_dict, (DataLoader, DataLoaderDict), create_loader_iters
```

Comment:
```suggestion
                self.loaders, self.loaders_iter_state_dict, DataLoader, create_loader_iters
```

This should not be applied to the state dict, since those don't have the sampler or dataset attributes I think.